### ISSUE TITLE
Future questions

### ISSUE THREAD

**Issue by soundprojects** [soundprojects/trainsim#1]
2 Questions about futures:

1. I use `read_input()` which locks to retrieve console input. When I drop the future in my main function, this lock is not dropped?

2. How can I cancel this future so the program will exit when the counter reaches the maximum count?

```
///Reads input from stdin
///
///Gets a line from stdin and checks if it is a "q"
///For now nothing else happens
///TODO: set our counter to the number entered
///
async fn read_input(){
    let stdin = io::stdin();
    for line in stdin.lock().lines(){
        println!("{}", line.as_ref().unwrap());

        if line.unwrap() == "q"
        {
            println!("Quitting input");
            return
        }
    }
}

//Tokio::main macro translates the main function back to a non-async function
// .await calls are transformed to block_on to make for easy coding
#[tokio::main]
async fn main() {
    let mut counter: i32 = 0;

    let input_task = tokio::spawn(async{
        read_input().await;
    });

    let task = tokio::spawn(async move {
        run_timer(&mut counter).await;
        println!("Count is: {}", counter);
    });

    //block on our task
    let _result = task.await.unwrap(); 
    drop(input_task)

}
```

**Comment by ruurdadema** [comment]
1. Tokio::spawn returns a `JoinHandle`, and A JoinHandle detaches the associated task when it is dropped, which means that there is no longer any handle to the task, and no way to join on it. But AFAIK it doesn't do any cancellation on that task.

([Source](https://docs.rs/tokio/latest/tokio/task/struct.JoinHandle.html))
 
btw. Dropping happens as a result of a struct going out-of-scope, only in very rare cases you would call drop manually.

2. You might want to use [abort()](https://docs.rs/tokio/latest/tokio/task/struct.JoinHandle.html#method.abort) for that, which I think cancels the task.

**Comment by soundprojects** [comment]
Tried `abort()` but according to documentation, Stdin lock causes the task to never unblock so abort does not do anything. 

Even though I do not block on `read_input()`, when the countdown is done, this function keeps the program from exiting.
I tried to search for a Rust way of killing a thread from the outside, this does not exist.
I thought of using a keep_going kind of variable but for the function to check the state of this variable, there has to be at least one line input for the function to get out of the first cycle and check our keep_going variable.

I think this has to do with Rust way of using its STD library, so I am going to let this one go and use a different UI or terminal framework to work with the user input.

**Comment by ruurdadema** [comment]
It's the same for all programming languages and has to do with how the OS works. To be honest, I've had the same painful experience with stdin and this is also the reason why I suggested going with egui (or another UI thing) right away.
Kudos for trying though!

**Comment by soundprojects** [comment]
std input libraries at the lowest level tend to block in a way that does not allow to use this in a way that is usable for us at this time
